src/es6-module-bundler.ts

```diff
@@ -0,0 +1,124 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import generate from 'babel-generator';
+import * as babel from 'babel-types';
+import {Document} from 'polymer-analyzer';
+
+import {getAnalysisDocument} from './analyzer-utils';
+import {AssignedBundle, BundleManifest} from './bundle-manifest';
+import {Bundler} from './bundler';
+import {BundledDocument} from './document-collection';
+import {Es6Rewriter, getBundleModuleExportName, getModuleExportNames, hasDefaultModuleExport} from './es6-module-utils';
+import {ensureLeadingDot, stripUrlFileSearchAndHash} from './url-utils';
+
+export class Es6ModuleBundler {
+  document: Document;
+
+  constructor(
+      public bundler: Bundler,
+      public assignedBundle: AssignedBundle,
+      public manifest: BundleManifest) {
+  }
+
+  async bundle(): Promise<BundledDocument> {
+    this.document = await this._prepareBundleDocument();
+    const baseUrl = this.document.parsedDocument.baseUrl;
+    const es6Rewriter =
+        new Es6Rewriter(this.bundler, this.manifest, this.assignedBundle);
+    const {code} = await es6Rewriter.rollup(
+        baseUrl, this.document.parsedDocument.contents);
+    this.document =
+        await this.bundler.analyzeContents(this.assignedBundle.url, code);
+    return {
+      ast: this.document.parsedDocument.ast,
+      content: this.document.parsedDocument.contents,
+      files: [...this.assignedBundle.bundle.files]
+    };
+  }
+
+  /**
+   * Generate a fresh document to bundle contents into.  If we're building a
+   * bundle which is based on an existing file, we should load that file and
+   * prepare it as the bundle document, otherwise we'll create a clean/empty
+   * JS document.
```

Comment:
Maybe show a short example of what a serialized version would like like, since it's a little hard to see that from the AST building code.

Reply:
Added in commit 675f266807886f9c7c3c7e9b4bb1fb7a8b5398b2

```ts
/**
 * Generate code containing import statements to all bundled modules and
 * export statements to re-export their namespaces and exports.
 *
 * Example: a bundle containing files `module-a.js` and `module-b.js` would
 * result in a prepareBundleModule result like:
 *
 *     import * as $moduleA from './module-a.js';
 *     import * as $moduleB from './module-b.js';
 *     import $moduleBDefault from './module-b.js';
 *     export {thing1, thing2} from './module-a.js';
 *     export {thing3} from './module-b.js';
 *     export {$moduleA, $moduleB, $moduleBDefault};
 */
```